src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```
@@ -0,0 +1,364 @@
+// information of the function body is passed through with the Function template
+// parameter.
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_f(const gsl_vector* const x,
+                                       void* function_object,
+                                       gsl_vector* const f_of_x) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  gsl_vector_set_with_std_array(f_of_x,
+                                fn->operator()(gsl_to_std_array<Dim>(x)));
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_df(const gsl_vector* const x,
+                                        void* function_object,
+                                        gsl_matrix* const jacobian) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  gsl_matrix_set_with_std_array(jacobian,
+                                fn->jacobian(gsl_to_std_array<Dim>(x)));
+
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_fdf(const gsl_vector* const x,
+                                         void* function_object,
+                                         gsl_vector* const f_of_x,
+                                         gsl_matrix* const jacobian) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  const std::array<double, Dim> x_as_std_array = gsl_to_std_array<Dim>(x);
+  gsl_vector_set_with_std_array(f_of_x, fn->operator()(x_as_std_array));
+  gsl_matrix_set_with_std_array(jacobian, fn->jacobian(x_as_std_array));
+  return GSL_SUCCESS;
+}
+}  // end namespace detail
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The exception thrown when gsl_multiroot fails to converge.
+ */
+class rootfinding_exception : public std::exception {
+  const char* what() const noexcept override { return "gsl_multiroot failed."; }
```

Are you sure you can't provide a better error message? This would not help me at all if I encountered it. Could you at least return the error code from GSL or something? Even better would be to translate the error code into a useful message.